File: src/main/java/org/goafabric/personservice/extensions/KafkaInterceptor.java
package org.goafabric.personservice.extensions;

import io.opentelemetry.api.trace.Span;
import io.opentelemetry.context.Context;
import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.ListTopicsOptions;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.header.Headers;
import org.jspecify.annotations.NonNull;
import org.slf4j.MDC;
import org.springframework.boot.actuate.endpoint.annotation.Endpoint;
import org.springframework.boot.actuate.endpoint.annotation.ReadOperation;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
import org.springframework.kafka.core.ConsumerFactory;
import org.springframework.kafka.core.KafkaAdmin;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.listener.DeadLetterPublishingRecoverer;
import org.springframework.kafka.listener.DefaultErrorHandler;
import org.springframework.kafka.listener.RecordInterceptor;
import org.springframework.util.backoff.FixedBackOff;

import java.nio.charset.StandardCharsets;
import java.util.Set;
import java.util.concurrent.ExecutionException;

@Configuration
@Endpoint(id = "topics")
@SuppressWarnings("java:S2638")
public class KafkaInterceptor {
    private final KafkaAdmin kafkaAdmin;

    public KafkaInterceptor(KafkaAdmin kafkaAdmin) {
        this.kafkaAdmin = kafkaAdmin;
    }

    @Bean
    public ConcurrentKafkaListenerContainerFactory<String, Object> kafkaListenerContainerFactory (
            ConsumerFactory<String, Object> consumerFactory,
            RecordInterceptor<String, Object> recordInterceptor
    ) {
        var factory = new ConcurrentKafkaListenerContainerFactory<String, Object>();
        factory.setConsumerFactory(consumerFactory);
        factory.setRecordInterceptor(recordInterceptor);
        return factory;
    }

    @Bean
    public RecordInterceptor<String, Object> recordInterceptor() {
        return new RecordInterceptor<>() {
            @Override
            public ConsumerRecord<String, Object> intercept(@NonNull ConsumerRecord<String, Object> consumerRecord, @NonNull Consumer<String, Object> consumer) {
                UserContext.setContext(getValue(consumerRecord.headers(), "X-TenantId"), getValue(consumerRecord.headers(), "X-OrganizationId"),
                        getValue(consumerRecord.headers(), "X-Auth-Request-Preferred-Username"), null);
                configureLogsAndTracing();
                return consumerRecord;
            }

            @Override
            public void afterRecord(@NonNull ConsumerRecord<String, Object> consumerRecord, @NonNull Consumer<String, Object> consumer) {
                afterCompletion();
            }
        };
    }

    @Bean
    public DefaultErrorHandler deadLetterErrorHandler(KafkaTemplate<String, Object> kafkaTemplate) {
        var recoverer = new DeadLetterPublishingRecoverer(
                kafkaTemplate, (record, exception) ->
                new TopicPartition(
                        record.topic() + ".DLT",
                        record.partition()
                )
        );

        var errorHandler = new DefaultErrorHandler(recoverer, new FixedBackOff(1000L, 3));
        errorHandler.addNotRetryableExceptions(IllegalStateException.class, IllegalArgumentException.class);

        return errorHandler;
    }

    @ReadOperation
    public Set<String> topics() throws ExecutionException, InterruptedException {
        try (AdminClient client = AdminClient.create(kafkaAdmin.getConfigurationProperties())) {
            return client.listTopics(new ListTopicsOptions().listInternal(false)).names().get();
        }
    }

    private void configureLogsAndTracing() {
        Span.fromContext(Context.current()).setAttribute("tenant.id", UserContext.getTenantId());
        MDC.put("tenantId", UserContext.getTenantId());
    }

    private void afterCompletion() {
        UserContext.removeContext();
        MDC.remove("tenantId");
    }

    private String getValue(Headers headers, String key) {
        return new String(headers.lastHeader(key).value(), StandardCharsets.UTF_8);
    }

}

File: src/main/kotlin/org/goafabric/personservice/extensions/KafkaInterceptor.kt
package org.goafabric.personservice.extensions

import io.opentelemetry.api.trace.Span
import io.opentelemetry.context.Context
import org.apache.kafka.clients.admin.AdminClient
import org.apache.kafka.clients.admin.ListTopicsOptions
import org.apache.kafka.clients.consumer.Consumer
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.common.header.Headers
import org.goafabric.personservice.extensions.UserContext.removeContext
import org.goafabric.personservice.extensions.UserContext.setContext
import org.goafabric.personservice.extensions.UserContext.tenantId
import org.slf4j.MDC
import org.springframework.boot.actuate.endpoint.annotation.Endpoint
import org.springframework.boot.actuate.endpoint.annotation.ReadOperation
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory
import org.springframework.kafka.core.ConsumerFactory
import org.springframework.kafka.core.KafkaAdmin
import org.springframework.kafka.listener.RecordInterceptor
import java.nio.charset.StandardCharsets

@Configuration
@Endpoint(id = "topics")
class KafkaInterceptor(private val kafkaAdmin: KafkaAdmin) {
    @Bean
    fun kafkaListenerContainerFactory(
        consumerFactory: ConsumerFactory<String, Any>,
        recordInterceptor: RecordInterceptor<String, Any>
    ): ConcurrentKafkaListenerContainerFactory<String, Any> {
        val factory = ConcurrentKafkaListenerContainerFactory<String, Any>()
        factory.setConsumerFactory(consumerFactory)
        factory.setRecordInterceptor(recordInterceptor)
        return factory
    }

    @Bean
    fun recordInterceptor(): RecordInterceptor<String, Any> {
        return object : RecordInterceptor<String, Any> {
            override fun intercept(
                consumerRecord: ConsumerRecord<String, Any>,
                consumer: Consumer<String, Any>
            ): ConsumerRecord<String, Any> {
                setContext(
                    getValue(consumerRecord.headers(), "X-TenantId"), getValue(consumerRecord.headers(), "X-OrganizationId"),
                    getValue(consumerRecord.headers(), "X-Auth-Request-Preferred-Username"), null
                )
                configureLogsAndTracing()
                return consumerRecord
            }

            override fun afterRecord(consumerRecord: ConsumerRecord<String, Any>, consumer: Consumer<String, Any>) {
                afterCompletion()
            }
        }
    }

    @ReadOperation
    fun topics(): MutableSet<String>? {
        AdminClient.create(kafkaAdmin.getConfigurationProperties()).use { client ->
            return client.listTopics(ListTopicsOptions().listInternal(false)).names().get()
        }
    }

    private fun configureLogsAndTracing() {
        Span.fromContext(Context.current()).setAttribute("tenant.id", tenantId)
        MDC.put("tenantId", tenantId)
    }

    private fun afterCompletion() {
        removeContext()
        MDC.remove("tenantId")
    }

    private fun getValue(headers: Headers, key: String): String {
        return String(headers.lastHeader(key).value(), StandardCharsets.UTF_8)
    }
}
